lib/SILOptimizer/Mandatory/MandatoryCombine.cpp

```diff
@@ -270,6 +271,25 @@ SILInstruction *MandatoryCombiner::visitApplyInst(ApplyInst *instruction) {
   return nullptr;
 }
 
+/// Try to remove partial applies that are no longer used
+SILInstruction *MandatoryCombiner::visitPartialApplyInst(PartialApplyInst *i) {
+  auto user = i->getSingleUse();
+  // If there are zero uses, this will be removed by dead code elimination.
+  // If there are more than one uses, we don't want to remove this.
+  if (!user)
+    return nullptr;
+  if (!isa<DeallocStackInst>(user->getUser()))
+    return nullptr;
+
+  // Remove the stack dealloc, then the partial apply, then the function ref.
+  instModCallbacks.deleteInst(user->getUser());
+  instModCallbacks.deleteInst(i);
+  // If the only use of the function_ref is us, then remove it.
+  if (i->getCallee()->getSingleUse()->getUser() == i)
+    instModCallbacks.deleteInst(cast<FunctionRefInst>(i->getCallee()));
```

Comment:
I don't think you can guarantee that the callee is always going to be a function_ref. Can you add an isa check to make sure?